[wizard] extend the wizard dashboard API to allow upload and empty config options

maximmaxim345 opened the pull request:
# What does this implement/fix?

This PR extends the `wizard` dashboard API to allow creating a new device with an empty or an uploaded configuration.

Required by https://github.com/esphome/dashboard/pull/782

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes https://github.com/esphome/backlog/issues/7

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5294

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

maximmaxim345 committed:
add a `minimal` wizard type

maximmaxim345 committed:
add a `upload` wizard type

maximmaxim345 committed:
remove invalid `type` variable

maximmaxim345 committed:
fallback to `basic` wizard type

maximmaxim345 committed:
Add new `type` parameter in tests

maximmaxim345 committed:
Add error message if upload file is invalid

maximmaxim345 committed:
Add tests for `empty` and `upload` wizard configs

jesserockz commented on esphome/wizard.py:
```diff
@@ -189,32 +191,49 @@ def wizard_write(path, **kwargs):
+    elif kwargs["type"] == "upload":
+        try:
+            file_text = base64.b64decode(kwargs["file_content"]).decode("utf-8")
+        except (binascii.Error, UnicodeDecodeError):
```
This is failing on characters like `°`

jesserockz replied:
We should probably also do this part in the webserver handler so we can propagate the error back to the user on the dashboard.

maximmaxim345 replied:
> This is failing on characters like `°`

I think you already fixed this with https://github.com/esphome/dashboard/pull/782/commits/78b223a2b247d5555a5fba422a0268ae7ae67520. At least it doesn't fail for me on files with ° and emojis anymore.

> We should probably also do this part in the webserver handler so we can propagate the error back to the user on the dashboard.

Done, also opened this PR:
- https://github.com/esphome/dashboard/pull/792

To show the specific error reason.

maximmaxim345 committed:
Move file decoding to `WizardRequestHandler`

To send the error message to the web client instead of logging it on the
server.

maximmaxim345 commented:
Noticed another issue while testing.
In case you create a new device configuration with an existing name, the old one just gets silently overwritten.
This affects both the existing new device wizard, but also uploading the same file twice.

Should we:
- Give an error?
- Silently change the name to something like "Desk Lamp (2)" if "Desk Lamp" already exists?
- Or specifically for this upload feature: Explicitly ask the user for a device name and not parse the file name as I done here?

maximmaxim345 committed:
Merge branch 'dev' into feat/wizard-add-upload-and-empty-config

jesserockz commented:
I'm fine with the upload feature using the uploaded filename, so in the case it already exists, we should throw an error back to the user.

maximmaxim345 committed:
Disallow overwriting existing config files with the wizard

maximmaxim345 committed:
Add test for overwrite protection

maximmaxim345 committed:
Show an error in case `write_wizard` fails

maximmaxim345 committed:
Add additional error message due to file conflict for the webserver

maximmaxim345 committed:
Merge branch 'dev' into feat/wizard-add-upload-and-empty-config
